### PR TITLE
Add slash commands to discord4j-rest and discord4j-gateway

### PR DESCRIPTION
**Pull request by UnderMybrella**
**Description:** This PR adds the necessary routes for creating and responding to slash commands.

**Justification:** #818 and [Discord API Changes](https://github.com/discord/discord-api-docs/pull/2295)

This PR depends on [Discord4J/discord-json#43](https://github.com/Discord4J/discord-json/pull/43).

This does **not** implement components in discord4j-core yet, as that will require a bit of extra work and design.

### CODE REVIEW

**Comment by UnderMybrella** [comment]
I should also mention that this includes a backport of webhook execution to 3.1.x which is a *little bit clunky*, but it's necessary for follow-up messages.

File: build.gradle
Comment (Shadorc) [inline build.gradle:63]: Is this change intended?

Comment (quanticc) [reply build.gradle:63]: I suggested this for testing, but once we merge Discord4J/discord-json#43 it won't be needed.

File: gradle.properties
Comment (quanticc) [inline gradle.properties:5]:
```suggestion
discordJsonVersion=1.5.7-SNAPSHOT
```

**Review by quanticc — CHANGES_REQUESTED** [review]
Please apply my suggestion to make it pass CI

File: rest/src/main/java/discord4j/rest/util/ApplicationCommandOptionType.java
Comment (j0rdanit0) [inline rest/src/main/java/discord4j/rest/util/ApplicationCommandOptionType.java:65]: Using a switch here requires the redundant association of the value with its enum type, which has already been defined in its constructor. Also, if a new type is added, this switch will have to be modified. Perhaps using a Stream would be a more scalable solution:

```java
return Arrays.stream(values()).filter(type -> type.getValue() == value).findFirst().orElse(UNKNOWN);
```

File: rest/src/main/java/discord4j/rest/util/InteractionResponseType.java
Comment (j0rdanit0) [inline rest/src/main/java/discord4j/rest/util/InteractionResponseType.java:73]: See my similar comment in ApplicationCommandOptionType.java

```java
return Arrays.stream(values()).filter(type -> type.getValue() == value).findFirst().orElse(UNKNOWN);
```

**Comment by quanticc** [comment]
A snapshot for this PR is available at `3.1.4-PR.823.0-SNAPSHOT` in case jitpack doesn't work

https://github.com/Discord4J/Discord4J/wiki/Using-latest-snapshots

File: rest/src/main/java/discord4j/rest/util/ApplicationCommandOptionType.java
Comment (UnderMybrella) [reply rest/src/main/java/discord4j/rest/util/ApplicationCommandOptionType.java:65]: I mirrored the existing enums for similar types like Channel.Type; I don't disagree with the proposal, but I think it's also important to be consistent.

Comment (quanticc) [reply rest/src/main/java/discord4j/rest/util/ApplicationCommandOptionType.java:65]: Please go ahead with the suggestion so this may drive improving existing enum code.

wdyt @darichey @danthonywalker

Comment (UnderMybrella) [reply rest/src/main/java/discord4j/rest/util/ApplicationCommandOptionType.java:48]: Gotcha! Included now

File: rest/src/main/java/discord4j/rest/service/ApplicationService.java
Comment (quanticc) [inline rest/src/main/java/discord4j/rest/service/ApplicationService.java:95]: Can you move these methods to InteractionService, so we can keep it local to this feature.

image

In the future we'll plan to resolve the inconsistencies, maybe by moving `/oauth2/applications` route to a dedicated OAuth2 class if we cover #715

**Review by quanticc — CHANGES_REQUESTED** [review]
A small request in this review.

I'm waiting for core member reviews otherwise we'll merge it tomorrow or Sunday.

File: rest/src/main/java/discord4j/rest/service/ApplicationService.java
Comment (quanticc) [reply rest/src/main/java/discord4j/rest/service/ApplicationService.java:95]: Actually, I'm not 100% on this. Implementations need to use that `/oauth2/applications` endpoint to get the application ID for endpoints, right?

If that's the case, I can now see the reasoning behind locating these methods here. What do you think?

**Review by quanticc — APPROVED** [review]

**Review by quanticc — CHANGES_REQUESTED** [review]
Noticed that we're missing these endpoints

PATCH /webhooks/{application.id}/{interaction.token}/messages/@original
DELETE /webhooks/{application.id}/{interaction.token}/messages/@original
POST /webhooks/{application.id}/{interaction.token}/messages
PATCH /webhooks/{application.id}/{interaction.token}/messages/{message.id}

**Review by quanticc — APPROVED** [review]
I'll apply WebhookService changes regarding String messageId

**Pull request merged**
